masonry/src/doc/02_implementing_widget.md

```diff
@@ -0,0 +1,306 @@
+
+We've also implemented the `on_access_event` method, which emulates the click behaviors for people using assistive technologies.
+
+Next we can leave the `on_anim_frame` and `update` implementations empty:
+
+```ignore
+use masonry::{
+    UpdateCtx
+};
+
+impl Widget for ColorRectangle {
+    // ...
+
+    fn on_anim_frame(&mut self, _ctx: &mut UpdateCtx, _interval: u64) {}
+    fn update(&mut self, _ctx: &mut UpdateCtx, _event: &Update) {}
+
+    // ...
+}
+```
+
+Next we implement layout:
+
+```ignore
+use masonry::{
+    LayoutCtx, BoxConstraints
+};
+
+impl Widget for ColorRectangle {
+    // ...
+
+    fn layout(&mut self, _ctx: &mut LayoutCtx, _bc: &BoxConstraints) -> Size {
+        self.size
+    }
+
+    // ...
+}
+```
+
+Our size is static, and doesn't depend on size constraints passed by our parent or context information like "the widget is currently hovered", so it can be written as a one-liner.
+
+Next we write our render methods:
+
+```ignore
+use masonry::{
+    PaintCtx, AccessCtx
+};
+use vello::Scene;
+use accesskit::{NodeBuilder, Role};
+
+impl Widget for ColorRectangle {
+    // ...
+
+    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
+        let rect = ctx.size().to_rect();
+        scene.fill(
+            Fill::NonZero,
+            Affine::IDENTITY,
+            self.color,
+            Some(Affine::IDENTITY),
+            &rect,
+        );
+    }
+
+    fn accessibility_role(&self) -> Role {
+        Role::Button
+    }
+
+    fn accessibility(&mut self, ctx: &mut AccessCtx, node: &mut NodeBuilder) {
+        node.set_default_action_verb(DefaultActionVerb::Click);
+    }
+
+    // ...
+}
+```
+
+In our `paint` method, we're given a `vello::Scene` and paint a rectangle into it.
+
+The rectangle's position is zero (because coordinates in our scenes are local to our widget), and its size is `ctx.size()`, which is the value returned by `layout()`; though we could also have used `self.size`.
+
+Next we define our accessibility role.
+Returning `Role::Button` means that screen readers will report our widget as a button, which roughly makes sense since it is clickable.
+
+<!-- TODO - Add more detail about how you should choose your role. -->
+
+In `accessibility`, we define a default action of `Click`, which is how we register our widget to be eligible for the `accesskit::Action::Default` event reported above.
+
+<!-- TODO - Is that actually true? I'm not sure what set_default_action does. -->
+
+Finally, we stub in some additional methods:
+
+```ignore
+use masonry::{
+    RegisterCtx, WidgetId
+};
+use smallvec::SmallVec;
+use tracing::{trace_span, Span};
+
+impl Widget for ColorRectangle {
+    // ...
+
+    fn register_children(&mut self, _ctx: &mut RegisterCtx) {}
+    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
+        SmallVec::new()
+    }
+
+    fn make_trace_span(&self) -> Span {
+        trace_span!("ColorRectangle")
+    }
+}
+```
+
+Don't worry about what they mean for now.
```

Comment:
I think here is the right place for `make_trace_span` to be described.

I'd probably also say something like:
```md
The children methods are described in [the next chapter](crate::doc::doc_03_implementing_container_widget).
Our widget can use the null implementations as it has no children:
```